File: 60-permutation_sequence.js
/**
 * Returns the k-th permutation sequence of numbers from 1 to n.
 * @param {number} n - The number of elements in the sequence.
 * @param {number} k - The index (1-based) of the permutation sequence to find.
 * @return {string} - The k-th permutation sequence as a string.
 */
var getPermutation = function(n, k) {
    // Generate a list of numbers from 1 to n
    let numbers = [];
    for (let i = 1; i <= n; i++) {
        numbers.push(i);
    }
    
    // Factorials array to store precomputed factorial values
    let fact = [1];
    for (let i = 1; i <= n; i++) {
        fact[i] = fact[i - 1] * i;
    }
    
    // Convert k to 0-based index
    k--;
    let result = "";
    
    // Construct the k-th permutation
    for (let i = n; i > 0; i--) {
        let index = Math.floor(k / fact[i - 1]); // Determine the index in the list
        result += numbers[index]; // Append the selected number
        numbers.splice(index, 1); // Remove used number
        k %= fact[i - 1]; // Update k for the next round
    }
    
    return result;
};

// Example usage:
console.log(getPermutation(3, 3)); // Output: "213"
console.log(getPermutation(4, 9)); // Output: "2314"

File: 60-permutation_sequence.py
from math import factorial

def getPermutation(n: int, k: int) -> str:
    # Create a list of numbers from 1 to n
    numbers = list(range(1, n + 1))
    
    # Convert k to zero-based index
    k -= 1
    result = []
    
    # Iterate to build the k-th permutation
    for i in range(n, 0, -1):
        # Determine the index of the current number
        index = k // factorial(i - 1)
        
        # Append the selected number to the result
        result.append(str(numbers[index]))
        
        # Remove the used number from the list
        numbers.pop(index)
        
        # Update k for the next selection
        k %= factorial(i - 1)
    
    return "".join(result)

# Example usage
print(getPermutation(3, 3))  # Output: "213"
print(getPermutation(4, 9))  # Output: "2314"
